src/main/scala/com/whitepages/cloudmanager/state/SolrState.scala

```diff
@@ -128,48 +180,104 @@ case class SolrState(state: ClusterState, collectionInfo: CollectionInfo, config
+    *                      an exact representation of the node in ZK
     * @param allowOfflineReferences
     * @throw ManagerException if a host could not be safely determined
     * @return A known canonical host
     */
   def canonicalNodeName(hostIndicator: String, allowOfflineReferences: Boolean = false): String = {
+    val rawNodeList = if (allowOfflineReferences) allNodes else liveNodes
 
-    def unambiguousFragment(fragment: String, dnsMap: Map[String,String]): Option[String] = {
-      findUnambigousNode(dnsMap, (s: String) => s == fragment)
-        .orElse(findUnambigousNode(dnsMap, (s: String) => s.contains(fragment)))
+    //If the value specified by the user exactly matches a node from the cluster state, return this value as is
+    if (rawNodeList.contains(hostIndicator)) {
+      hostIndicator
     }
-    def findUnambigousNode(dnsMap: Map[String,String], comparison: (String) => Boolean): Option[String] = {
-      val matchingMaps = dnsMap.filter{
-        case (dnsName,canonName) => comparison(dnsName) || comparison(canonName)
-      }
-      matchingMaps.toList match {
-        case (dnsName, canonName) :: Nil => Some(canonName)
-        case _ => None
-      }
+    else {
+      unambiguousFragment(hostIndicator, dnsNameMap(rawNodeList)).getOrElse(
+        unambiguousFragment(hostIndicator, ipNameMap(rawNodeList)).getOrElse(
+          {
+            //Here we attempt to transform the indicator(user passed in value) instead of the values from the cluster state
+            val chunks = hostIndicator.split(':')
+            val host = chunks.head
+            val port = if (chunks.length > 1) ":" + chunks.last else ""
+            val ipAndPort = InetAddress.getByName(host).getHostAddress + port
+            val matches = rawNodeList.filter((node) => node.contains(ipAndPort))
+            matches.size match {
+              case 0 => throw new ManagerException(s"Could not determine a live node from '$hostIndicator'")
+              case 1 => matches.head
+              case _ => throw new ManagerException(s"Ambiguous node name '$hostIndicator', possible matches: $matches")
+            }
+          }
+        )
+      )
     }
+  }
 
-    val nodeList = if (allowOfflineReferences) allNodes else liveNodes
-    if (nodeList.contains(hostIndicator)) {
-      hostIndicator
+
+  /**
+    *
+    * @param fragment          user passed in string for node identification
+    * @param nodeComparisonMap map where key is the resolved name of a node(IP or Host) and value is its
+    *                          string representation in the cluster state
+    * @return
+    */
+  def unambiguousFragment(fragment: String, nodeComparisonMap: Map[String,String]): Option[String] = {
+    findUnambigousNode(nodeComparisonMap, (s: String) => s == fragment)
+      .orElse(findUnambigousNode(nodeComparisonMap, (s: String) => s.contains(fragment)))
```

Comment:
I think we should get rid of the .contains, it prohibits exact IP/Hostname specification. I believe this happens with the existing version of solr cloud manager also. 
For e.g specifying --nodes 1.0.0.8 selects nodes 1.0.0.8,1.0.0.81 and 1.0.0.82

Reply:
So the ambiguous node exception will not be thrown every time, IMO it should be. For e.g. if you invoke waitactive with an IP that matches multiple nodes, the call just sums up the replica count across those nodes and returns their state e.g. "All 18 shards were active". IMO we should enforce really tight matching in this case.